packages/web-components/src/components/fade-in-out/fade-in-out.ts

```diff
@@ -0,0 +1,256 @@
+  }
+
+  /**
+   * Cleans observers upon update or exit, and creates new instances if needed.
+   *
+   * @param [options] The options.
+   * @param [options.create] `true` to create the new intersection observer.
+   */
+  private _cleanAndCreateObservers({ create }: { create?: boolean } = {}) {
+    this._cleanAndCreateRootObserver();
+    this._cleanAndCreateInnerObserver();
+    this._cleanAndCreateResizeObserver();
+
+    if (create) {
+      this._cleanAndCreateRootObserver({ create: true });
+      this._cleanAndCreateResizeObserver({ create: true });
+
+      const { selectorTargets } = this;
+      if (selectorTargets) {
+        forEach(this.ownerDocument!.querySelectorAll(selectorTargets), item => {
+          this._rootObserver?.observe(item);
+        });
+      }
+      this._resizeObserver.observe(this.ownerDocument!.documentElement);
+    }
+  }
+
+  /**
+   * Cleans-up and creats the root intersection observer.
+   *
+   * @param [options] The options.
+   * @param [options.create] `true` to create the new intersection observer.
+   */
+  private _cleanAndCreateRootObserver({ create }: { create?: boolean } = {}) {
+    if (this._rootObserver) {
+      this._rootObserver.disconnect();
+      this._rootObserver = null;
+    }
+
+    if (create) {
+      this._rootObserver = new IntersectionObserver(this._handleExit);
+    }
+  }
+
+  /**
+   * Cleans-up and creats the inner intersection observer.
+   *
+   * @param [options] The options.
+   * @param [options.create] `true` to create the new intersection observer.
+   */
+  private _cleanAndCreateInnerObserver({ create }: { create?: boolean } = {}) {
+    if (this._innerObserver) {
+      this._innerObserver.disconnect();
+      this._innerObserver = null;
+    }
+
+    if (create) {
+      this._innerObserver = new IntersectionObserver(this._handleEntrance.bind(this), { rootMargin: this._rootMargin });
+      const { selectorTargets } = this;
+      if (selectorTargets) {
+        forEach(this.ownerDocument!.querySelectorAll(selectorTargets), item => {
+          this._innerObserver?.observe(item);
+        });
+      }
+    }
+  }
+
+  /**
+   * Cleans-up and creats the resize observer.
+   *
+   * @param [options] The options.
+   * @param [options.create] `true` to create the new resize observer.
+   */
+  private _cleanAndCreateResizeObserver({ create }: { create?: boolean } = {}) {
+    if (this._resizeObserver) {
+      this._resizeObserver.disconnect();
+      this._resizeObserver = null;
+    }
+
+    if (create) {
+      // TODO: Wait for `.d.ts` update to support `ResizeObserver`
+      // @ts-ignore
+      this._resizeObserver = new ResizeObserver(this.handleResize.bind(this));
+    }
+  }
+
+  /**
+   * Handler to add recalculated rootMargin to observer.
+   *
+   * @private
+   */
+  private handleResize() {
+    this._rootMargin = this._getViewportMargin();
+    this._cleanAndCreateInnerObserver({ create: true });
```

Comment:
Can we remove `_getViewportMargin()` private method and send the calculated viewport margin to `._cleanAndCreateInnerObserver()` like `._cleanAndCreateInnerObserver({ create: true, viewportMargin: `-${((this.ownerDocument!.documentElement.clientHeight * _colSpan) / breakpoints.max.columns).toString()}px 0px` })`? Wanted to ensure viewport margin calculation won't happen except in `_handleResize()`.